capsules/extra/src/distance.rs

```diff
@@ -0,0 +1,195 @@
+//! * `ENOSUPPORT`: Invalid `cmd`.
+//! * `NOMEM`: Insufficient memory available.
+//! * `INVAL`: Invalid address of the buffer or other error.
+//!
+//! Usage
+//! -----
+//!
+//! You need a device that provides the `hil::sensors::Distance` trait.
+//! Here is an example of how to set up a distance sensor with the HC-SR04.
+//!
+//! ```rust,ignore
+//! # use kernel::static_init;
+//!
+//! let trig_pin = peripherals.pins.get_pin();
+//! let echo_pin = peripherals.pins.get_pin();
+//! echo_pin.make_input();
+//! trig_pin.make_output();
+//!
+//! let virtual_alarm_hc_sr04 = static_init!(
+//!     VirtualMuxAlarm<'static, RPTimer>,
+//!     VirtualMuxAlarm::new(mux_alarm)
+//! );
+//! virtual_alarm_hc_sr04.setup();
+//!
+//! let hc_sr04 = static_init!(
+//!     hc_sr04::HcSr04<VirtualMuxAlarm<'static, RPTimer>>,
+//!     hc_sr04::HcSr04::new(trig_pin, echo_pin, virtual_alarm_hc_sr04)
+//! );
+//! virtual_alarm_hc_sr04.set_alarm_client(hc_sr04);
+//!
+//! echo_pin.set_client(hc_sr04);
+//!
+//! let distance_sensor = static_init!(
+//!     distance::DistanceSensor<
+//!         'static,
+//!         hc_sr04::HcSr04<'static, VirtualMuxAlarm<'static, RPTimer>>,
+//!     >,
+//!     distance::DistanceSensor::new(
+//!         hc_sr04,
+//!         board_kernel.create_grant(distance::DRIVER_NUM, &memory_allocation_capability)
+//!     )
+//! );
+//! hc_sr04.set_client(distance_sensor);
+//! ```
+
+use core::cell::Cell;
+
+use kernel::grant::{AllowRoCount, AllowRwCount, Grant, UpcallCount};
+use kernel::hil;
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::{ErrorCode, ProcessId};
+
+/// Syscall driver number.
+use capsules_core::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Distance as usize;
+
+#[derive(Default)]
+pub struct App {
+    subscribed: bool,
+}
+
+pub struct DistanceSensor<'a, T: hil::sensors::Distance<'a>> {
+    driver: &'a T,
+    apps: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    busy: Cell<bool>,
+}
+
+impl<'a, T: hil::sensors::Distance<'a>> DistanceSensor<'a, T> {
+    pub fn new(
+        driver: &'a T,
+        grant: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    ) -> DistanceSensor<'a, T> {
+        DistanceSensor {
+            driver,
+            apps: grant,
+            busy: Cell::new(false),
+        }
+    }
+
+    fn enqueue_command(&self, processid: ProcessId) -> CommandReturn {
+        self.apps
+            .enter(processid, |app, _| {
+                // Unconditionally mark this client as subscribed so it will get
+                // a callback when we get the distance reading.
+                app.subscribed = true;
+
+                // If we do not already have an ongoing read, start one now.
+                if !self.busy.get() {
+                    self.busy.set(true);
+                    match self.driver.read_distance() {
+                        Ok(()) => CommandReturn::success(),
+                        Err(e) => CommandReturn::failure(e),
```

Comment:
If there's an error, don't we also need to clear `self.busy`?

Reply:
Maybe `app.subscribed` should also be set to false too? Otherwise you could later receive unexpected measurement callbacks.

Reply:
You're right.